lib/magma/validation/attribute.rb

```diff
@@ -0,0 +1,116 @@
+            yield error
+          end
+        end
+      end
+
+      private
+
+      def attribute_validations(att_name)
+        @attribute_validations[att_name] ||= validation(@model.attributes[att_name]).new(
+          @model, @validator, @model.attributes[att_name]
+        )
+      end
+
+      def validation(att)
+        validation_class = :"#{att.class.to_s.split('::').last}Validation"
+        Magma::Validation::Attribute.const_defined?(validation_class) ?
+          Magma::Validation::Attribute.const_get(validation_class) :
+          Magma::Validation::Attribute::BaseAttributeValidation
+      end
+
+      class BaseAttributeValidation
+        def initialize(model, validator, attribute)
+          @model = model
+          @validator = validator
+          @attribute = attribute
+        end
+
+        def validate(value)
+          # do nothing
+        end
+      end
+
+      class AttributeValidation < BaseAttributeValidation
+        def validate(value)
+          case match
+          when Regexp
+            yield format_error(value) if !match.match(value)
+          when Array
+            if !match.map(&:to_s).include?(value)
+              yield "On #{@attribute.name}, '#{value}' should be one of #{match.join(", ")}."
+            end
+          end
+        end
+
+        private
+
+        def match
+          @match ||= @attribute.match.is_a?(Proc) ? @attribute.match.call : @attribute.match
+        end
+
+        def format_error(value)
+          if @attribute.format_hint
+            "On #{@attribute.name}, '#{value}' should be like '#{@attribute.format_hint}'."
+          else
+            "On #{@attribute.name}, '#{value}' is improperly formatted."
+          end
+        end
+      end
+
+      class ChildValidation < BaseAttributeValidation
+        def validate(value)
+          return if value.nil? || value.empty?
+          @validator.validate(@attribute.link_model, @attribute.link_model.identity => value) do |error|
+            yield error
+          end
+        end
+      end
+      class ForiegnKeyValidation < BaseAttributeValidation
+        def validate(value)
+          return if value.is_a?(Magma::TempId) || value.nil?
+          if @attribute.link_identity
+            @validator.validate(@attribute.link_model, @attribute.link_model.identity => value) do |error|
+              yield error
+            end
+          end
+        end
+      end
+      class CollectionValidation < BaseAttributeValidation
+        def validate(value)
+          unless value.is_a?(Array)
+            yield "#{value} is not an Array."
+            return
+          end
+          value.each do |link|
+            next unless link
+            @validator.validate(@attribute.link_model, @attribute.link_model.identity => link) do |error|
+              yield error
+            end
+          end
+        end
+      end
+      class TableValidation < CollectionValidation; end
```

Comment:
What is TableValidation for?

Reply:
I couldn't decide where I wanted the attribute validation code to live. I ended up putting it back into the attribute class (i.e. Magma::CollectionAttribute::Validation < Magma::Validation::Attribute::BaseAttributeValidation). The same line with respect to TableAttributes becomes: `class Validation < Magma::CollectionAttribute::Validation; end` These attribute validation classes are found (if named `Validation`) by Magma::Validation::Attribute#validate on the corresponding attribute object's class. I wanted the TableAttribute to duplicate the validation behavior of the CollectionAttribute. Therefore, this class Magma::TableAttribute::Validation had to exist and inherit from CollectionAttribute. In the future it might have its own additional behavior.

 Not sure this is the best way to accomplish this organization, but it seemed to make the most sense of the available options. This is also the way the loader's Entry interface currently works (with a class Magma::WhateverAttribute::Entry) to define attribute-specific database entries, so these two organization mechanisms are at least consistent.